### PR TITLE
Service worker

### PR DESCRIPTION
**Pull request by rosemcc**
This PR adds PWA functionality to the ResearchHub, using the [Angular service worker](https://angular.io/guide/service-worker-intro). A service worker is a script that runs in the web browser and manages caching for an application, by acting as a network proxy. Using a service worker to reduce dependency on the network can significantly improve reliability and performance of the app. In addition to caching improvements, having the app configured as a PWA also allows it to be installable on the user's device/platform, which can increase user engagement.

Service workers are supported in the latest versions of Chrome, Firefox, Edge, Safari, Opera, UC Browser (Android version) and Samsung Internet. IE and Opera Mini do not support service workers. Where service workers are not supported, the app will still run as normal, but without the added benefits of a service worker.

## Proposed Changes

* Addition of the following files to support service worker and PWA functionality:
  * ngsw-config.json - Angular service worker configuration. This file specifies the caching behaviours of the service worker including which file types and external resources to cache, and the [install mode](https://angular.io/guide/service-worker-config#installmode) and [update mode](https://angular.io/guide/service-worker-config#updatemode) to use for different resources. It also configures caching policies for data (API) requests. We specify a policy for the graphql service with the ['freshness' strategy ](https://angular.io/guide/service-worker-config#strategy) which means the service worker will preferentially fetch requested data from the network. Only if the network times out, (according to 'timeout' setting), does the request fall back to the cache. This makes sense for our purposes, as the app content could be updated at any time, so we ensure that the data displayed is always up to date with the CMS.
  * manifest.webmanifest - Web app manifest which configures the app name, theme color, preferred display settings, scope and start url, and specifies the location of app icons. This is the file which enables the app to be installed on a users device.
  * assets/icons - Series of app icons used for display on notifications and app installations, with different sizes to accommodate different devices/OS.

* Updated app.module.ts to register and enable the service worker script, and added a bypass for 504 (gateway timeout) errors so that the service worker can attempt to serve cached files if there is a problem with the network connection (enable offline functionality of the app)
* Updated angular.json to include the app manifest in the build
* App manifest link and app theme color is added to index.html
* Updated README
* Added script to run the app in prod mode to enable testing service worker functionality locally (`npm start:prod`)
* Update cloudfront secure-headers lambda function to include the app manifest, and modifications to connect-src policy for service worker requests ([more info](https://qubyte.codes/blog/content-security-policy-and-service-workers)).
* Added sw-updates.service.ts - manages updates to the app and user notifications, and handles error states, using the [SwUpdate](https://angular.io/api/service-worker/SwUpdate) class.
* Users are notified when a new version of the app is ready to install. The notification is an Angular Material [snackbar ](https://material.angular.io/components/snack-bar/overview)

## Additional changes/fixes

* Cleaned up codeCoverageExclude array in angular.json
* Fixed the 'about' link on the footer and fixed the unit test
* Removed many unused Material modules from app.material.module.ts

### CODE REVIEW

**Comment by Trombach** [comment]
This comment is pretty much just an FYI.

One minor issue we probably don't have to address can be viewed in the chrome dev tools under `Application` > `Manifest`
image

This is related to how a PWA is installable and certain platforms allow you to change the shape of the app launch icon. For those cases the provided icons have to be maskable https://web.dev/maskable-icon/

Given that the install-ability of the hub is only a side effect of this PR, it's not something we have to address. However, if someone decides to install the app there is a (low) chance that the icon will not look quite right.

File: research-hub-web/src/index.html [hunk 1]
```diff
@@ -12,21 +12,25 @@
...
 
     <!-- Twitter Card -->
-    <meta name="twitter:card" content="summary" />
-    <meta name="twitter:title" content="ResearchHub" />
-    <meta name="twitter:description" content="The ResearchHub connects you with people, resources, and services from across the University to enhance and accelerate your research." />
-    <meta name="twitter:image" content="https://images.ctfassets.net/vbuxn5csp0ik/5IfxoF4WQ4Ks71q2HeLbgC/d0c3bb826ceebd31be789a73edac48fe/UoA_Details_230-min.jpeg?w=120&h=120&fm=png&fit=thumb&f=face" />
+    <meta name="twitter:card" content="summary"/>
+    <meta name="twitter:title" content="ResearchHub"/>
+    <meta name="twitter:description" content="The ResearchHub connects you with people, resources, and services from across the University to enhance and accelerate your research."/>
+    <meta name="twitter:image" content="https://images.ctfassets.net/vbuxn5csp0ik/5IfxoF4WQ4Ks71q2HeLbgC/d0c3bb826ceebd31be789a73edac48fe/UoA_Details_230-min.jpeg?w=120&amp;h=120&amp;fm=png&amp;fit=thumb&amp;f=face"/>
```
Comment (Trombach) [inline research-hub-web/src/index.html:18]: Just out of curiosity, what are those `amp;`s for?

**Review by Trombach — CHANGES_REQUESTED** [review]
LGTM! Just some comments and one fix for the sw-update unit test.

Really cool stuff in here. The app looks awesome when installed and the update notification came through instantly. Nice work!

File: research-hub-web/src/app/services/sw-updates.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    expect(service.snackBarRef.instance.data.message).toBe(message);
+    expect(service.snackBarRef.instance.data.action).toBe(action);
+
+    
+    service.snackBarRef.afterDismissed().subscribe(() => {
+      expect(activateUpdateSpy).toHaveBeenCalled();
+    });
+
+    service.snackBarRef.dismissWithAction();
+  });
```
[the remark this answers is not in the document]
Comment (Trombach) [reply research-hub-web/src/app/services/sw-updates.service.spec.ts:44]: I've just tested this locally and with those changes the test fails. That means the current expect statement at line 40 doesn't actually get evaluated. We can have a chat if you want me to explain what's going on :)

File: research-hub-web/src/index.html (hunk 1 above)
Comment (rosemcc) [reply research-hub-web/src/index.html:18]: Good question... I'm trying to remember back to when I added that in but I can't remember doing anything specific with the url format?? It does seem to work fine when you post the ResearchHub link on Twitter:
image

**Pull request merged**